### ISSUE TITLE
Problems after compiling to EXE with pyinstaller and numba/llvmlite v0.61.0rc1/v0.44.0rc1 within CPython v3.13

### ISSUE THREAD

**Issue by Alpengreis** [numba/numba#9844]
I use the combo numba/llvmlite v0.61.0rc1/v0.44.0rc1 within CPython v3.13.

After I compiled a script to EXE with pyinstaller I received the following error:

`ModuleNotFoundError: No module named 'numba.core.types.old_scalars'`
 ... then ...
`ModuleNotFoundError: No module named 'numba.core.datamodel.old_models'`

and so far.

I could solve that with the "--hidden-import=" of pyinstaller, but that gave a long list:

`--hidden-import=numba.core.types.old_scalars --hidden-import=numba.core.datamodel.old_models --hidden-import=numba.cpython.old_builtins --hidden-import=numba.core.typing.old_builtins --hidden-import=numba.core.typing.old_cmathdecl --hidden-import=numba.core.typing.old_mathdecl --hidden-import=numba.cpython.old_hashing --hidden-import=numba.cpython.old_numbers --hidden-import=numba.cpython.old_tupleobj --hidden-import=numba.np.old_arraymath --hidden-import=numba.np.random.old_distributions --hidden-import=numba.np.random.old_random_methods --hidden-import=numba.cpython.old_mathimpl --hidden-import=numba.core.old_boxing`

THEN I could run the EXE.

I think this can be automated yet?

Greetings
Alpengreis

**Comment by gmarkall** [comment]
Does pyinstaller scan source files to look for import statements to determine what to bundle? I don't know the solution to your problem, but if that's the case, then I think it will be failing to detect redirected imports, like:

https://github.com/numba/numba/blob/8348b5d844d3f67ce5dc41d68a09a10dd5b6b927/numba/core/datamodel/models.py#L1-L12

that use `_RedirectSubpackage`:

https://github.com/numba/numba/blob/8348b5d844d3f67ce5dc41d68a09a10dd5b6b927/numba/core/utils.py#L674-L686

Can pyinstaller be configured with additional logic to determine imports? If so, perhaps it could inspect the uses of `_RedirectSubpackage`?

**Comment by Alpengreis** [comment]
But why does it work then without any problems with CPython 3.12.7 and current release of numba?

**Comment by Alpengreis** [comment]
Can nobody help here?

**Comment by gmarkall** [comment]
> But why does it work then without any problems with CPython 3.12.7 and current release of numba?

Because the `old_` / `new_` split and redirection mechanism was only introduced with the current RC, and was not in the current stable release of Numba.

**Comment by Alpengreis** [comment]
@gmarkall 

Ok, thanks for info. And will this split be integrated in the in the next stable release of Numba too? Hopefully not, else it would be VERY complicated to use Numba with "pyinstaller-to-EXE" ...

**Comment by gmarkall** [comment]
Can you open a PyInstaller issue to look into this issue? It's not clear how we can remedy this from the Numba side, so any additional input might be helpful to find a resolution.

**Comment by Alpengreis** [comment]
Will try, not sure about yet because it seems it's not an issue with PyInstaller ...

**Comment by Alpengreis** [comment]
Reported as following issue in Pyinstaller:

https://github.com/pyinstaller/pyinstaller/issues/8989

**Comment by Alpengreis** [comment]
I received an answer already!

**Comment by Alpengreis** [comment]
In the PyInstaller issue, they asked me the following:

> Do you know if this old vs new duality is a permanent or transitional thing? And is there any assurance you can give that numba won't come up with newer mode in the future and be forced to rename everything to something with version numbers in it to avoid having to call the newer mode newnew?

How is that?

**Comment by Alpengreis** [comment]
Seems fixed with next PyInstaller HooksContrib ...

**Comment by Alpengreis** [comment]
Yes, with update to "pyinstaller_hooks_contrib-2025.1" it's fixed.

**Comment by esc** [comment]
@Alpengreis awesome, thank you for reporting back and closing this issue.
